### PR TITLE
Implement `Negate` (two's complement)

### PR DESCRIPTION
**Pull request by anurudhp**

### CODE REVIEW

File: qualtran/bloqs/arithmetic/negate_test.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+def test_examples(bloq_autotester):
+    bloq_autotester(_negate)
+    bloq_autotester(_negate_symb)
+
+
+@pytest.mark.parametrize("bitsize", [1, 2, 3, 4, 8])
+def test_negate_classical_sim(bitsize: int):
+    # TODO use QInt once classical sim is fixed.
+    #      classical sim currently only supports unsigned.
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/negate_test.py:28]: Is this true? I thought classical simulation supports both signed and unsigned integers. Is there an open issue you are referring to?

**Review by tanujkhattar — COMMENTED** [review]
Looking good. Question about classical sim currently only supportin unsigned types - is this true?

File: qualtran/bloqs/arithmetic/negate.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    the T-cost compared to a controlled adder.
+
+    Args:
+        dtype: The data type of the input value.
+
+    Registers:
+        x: Any signed value stored in two's complement form.
+    """
+
+    dtype: QDType
```
Comment (tanujkhattar) [inline qualtran/bloqs/arithmetic/negate.py:47]: Should we assert that `dtype` is a signed type ? So `QInt` or `QFxp(signed=True)` ?

File: qualtran/bloqs/mod_arithmetic/_shims.py [hunk 3]
```diff
@@ -159,7 +159,7 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         # return {(Toffoli(), 32 * self.n**2 * log2(self.n))}
         return {
             (_ModInvInner(n=self.n, mod=self.mod), 2 * self.n),
-            (Negate(self.n), 1),
+            (Negate(QUInt(self.n)), 1),
```
Comment (tanujkhattar) [inline qualtran/bloqs/mod_arithmetic/_shims.py:162]: What does this mean? Negating an unsigned type?

File: qualtran/bloqs/arithmetic/negate_test.py (hunk 1 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/negate_test.py:28]: https://github.com/quantumlib/Qualtran/issues/606

Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/negate_test.py:28]: Thanks, looks like it's bottlenecked by speed and https://github.com/quantumlib/Qualtran/pull/1142/ should help fix it?

File: qualtran/bloqs/arithmetic/negate.py (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/negate.py:47]: It actually works for unsigned types as well, it's the [unary minus operator in c++](https://en.cppreference.com/w/cpp/language/operator_arithmetic). Perhaps we can use the c++ naming convention?

File: qualtran/bloqs/mod_arithmetic/_shims.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/mod_arithmetic/_shims.py:162]: Ah I'm actually not sure where the costs are coming from. @mpharrigan I think you added this?

File: qualtran/bloqs/arithmetic/negate.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/negate.py:47]: Thanks, we should at least have a link explaining the behavior for unsigned types in the docstring. I don't have strong opinions on naming convention.

**Review by tanujkhattar — APPROVED** [review]
LGTM. Please consider adding a bitwise `Not` Bloq for `OnEach(XGate(), self.bitsize)` to `arithmetic/bitwise.py`. This can be done in a follow up PR so we can merge this.

File: qualtran/bloqs/arithmetic/negate.py (hunk 2 above)
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/negate.py:56]: updated docstring to explain this

Comment (anurudhp) [reply qualtran/bloqs/arithmetic/negate.py:56]: will leave the Bloq as `Negate` for now

File: qualtran/bloqs/arithmetic/negate.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+from qualtran import Bloq, bloq_example, BloqDocSpec, QDType, QInt, Signature
+from qualtran.bloqs.arithmetic import AddK
+from qualtran.bloqs.basic_gates import OnEach, XGate
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+
+
+@frozen
+class Negate(Bloq):
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/arithmetic/negate.py:28]: yep this is useful! will add in a follow-up PR and cleanup this decomp then.

File: qualtran/bloqs/mod_arithmetic/_shims.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/bloqs/mod_arithmetic/_shims.py:162]: I think you've already answered my question about what does it mean to negate an unsigned type.  I am not worried about the exact costs here. I'll resolve the comment.

**Pull request merged**